riff-raff/app/deployment/preview/Preview.scala

```diff
@@ -0,0 +1,45 @@
+package deployment.preview
+
+import cats.data.Validated
+import cats.data.Validated.{Invalid, Valid}
+import cats.kernel.Semigroup
+import cats.syntax.semigroup._
+import com.gu.management.Loggable
+import magenta.`package`._
+import magenta.artifact.S3YamlArtifact
+import magenta.deployment_type.DeploymentType
+import magenta.graph.{DeploymentTasks, Graph}
+import magenta.input.resolver._
+import magenta.input.{ConfigErrors, Deployment}
+import magenta.{DeployParameters, DeploymentResources}
+
+object Preview extends Loggable {
+  def apply(artifact: S3YamlArtifact, yamlConfig: String, parameters: DeployParameters,
+    resources: DeploymentResources, deploymentTypes: Seq[DeploymentType]): Preview = {
+
+    val validatedGraph = for {
+      deploymentGraph <- Resolver.resolveDeploymentGraph(yamlConfig, deploymentTypes)
+      flattenedGraph = DeploymentGraphActionFlattening.flattenActions(deploymentGraph)
+      previewGraph <- sequenceGraph {
+        flattenedGraph.map { deployment =>
+          TaskResolver.resolve(deployment, resources, parameters, deploymentTypes, artifact).map(deployment -> _)
+        }
+      }
+    } yield previewGraph
+    Preview(validatedGraph)
+  }
+
+  private[preview] def sequenceGraph[A, E](graph: Graph[Validated[E, A]])(implicit E: Semigroup[E]): Validated[E, Graph[A]] = {
```

Comment:
Might it make sense to implement `Traverse` (http://typelevel.org/cats/typeclasses/traverse.html) for `Graph`?